### PR TITLE
Improve property management of pools, add property management to datasets

### PR DESCRIPTION
**Pull request by MaienM**
This adds a simple map on the dataset that is synced with the local properties of the dataset (that is, the properties that are different from its parent).

### CODE REVIEW

**Comment by MathiasPius** [comment]
Hi Maien, first of all thanks for the contribution!

This feature should definitely be implemented, but there are a few obstacles. The way terraform handles state changes means that we have to deal with *desired* properties and *actual* properties a strangely, or at least that was my conclusion when I implemented it.

In the `zfs_pool` resource, I handle this by having two separate sets of properties, there's `property` which is a block structure used like so:

```terraform
resource "zfs_pool" "zdata" {
  name = "testpool"

  property {
    name = "autotrim"
    value = "on"
  }

  // ...
}```

And then there's an read-only variable called `properties` which I populate with the actual data. The reason for this is that if I used `properties` for both, the state would change immediately after application because it had retrieved new information after application from the pool's default values.

I could have simply ignored properties not explicitly listed in `properties`, but I like the ability to read properties from the terraform state of objects.

At this point I was going to point you at the code showing how I did this in the `pool` resource, but after having a look myself I realized that while it appears to be implemented (property and properties are both there), the actual `property`fields are never applied...

Give me some time to fix the property application for zpools, then I'll have something to point at as an example :)

**Comment by MaienM** [comment]
Good point about the full read-only list, hadn't considered that. I've updated the PR to use the attribute `local_properties` for the overrides, and to include the read-only attribute `properties` that contains all properties including default/inherited ones.

> The reason for this is that if I used properties for both, the state would change immediately after application because it had retrieved new information after application from the pool's default values.

I specifically check for the property's source being `local` to avoid default values from interfering. I've only just taken a quick glance at the code for pools, but it doesn't look like you make that distinction there, perhaps that's why you previously had issues? I consider the map of properties set in the resource to be the full list of desired local properties, and the properties with source `local` fetched from the dataset the actual list. As far as I could tell from my testing this worked in all cases; there were no spurious diffs caused by default values or anything of that sort.

As an aside, I hadn't seen the property management on the pool before, but I wonder why you've gone for a block per property rather than just a map. There are no more attributes to a property than its name and value, so this feels needlessly verbose. Was this just a matter of wanting to not use the `properties` attribute for this? I prefer the simpler scheme of just having map (for which I've now used `local_properties` to avoid any conflicts), but its not a big deal either way.

I wouldn't mind updating/fixing the pool code as part of this PR, the way to get/set properties is identical between pools and datasets so most of the code could be shared anyway. No need for us to separately spend time implementing this twice! Just let me know which flavor (`property` block vs `local_properties` map) you want to use and I'll update this PR so that both pools and datasets work.

**Comment by MaienM** [comment]
One caveat I just realized is that some automated tools use custom ZFS properties to track additional metadata, and we probably don't want to touch these. Using `property` blocks would allow us to mark a property as ignored/unmanaged, which the map wouldn't, so that's a reason to go with that over a simple map.

**Comment by MathiasPius** [comment]
It's been a while since I really dug into terraform provider structure, but like you said I think the reasoning for using the blocks was that you could track individual pieces of information without invalidating the whole, or it might have been an experiment in circumventing the desired/actual property issue which I never got around to rolling back.

I just finished fixing the pool options functionality, which I think provides a decent solution that can be applied to datasets as well, see the following links for reference:

**Schema**
https://github.com/MathiasPius/terraform-provider-zfs/blob/main/internal/provider/resource_pool.go#L89-L101

**Setting properties upon create**
https://github.com/MathiasPius/terraform-provider-zfs/blob/main/internal/provider/resource_pool.go#L142C4-L148
https://github.com/MathiasPius/terraform-provider-zfs/blob/main/internal/provider/zfs.go#L297-L300

**Modifying properties later**
https://github.com/MathiasPius/terraform-provider-zfs/blob/main/internal/provider/resource_pool.go#L235-L257
https://github.com/MathiasPius/terraform-provider-zfs/blob/main/internal/provider/zfs.go#L21-L37

Let me know what you think. I'm not opposed to rewriting it into a map, if it makes more sense, but I'd prefer to keep the same syntax as before, and for datasets and pools to share syntax regarding properties.

**Comment by MaienM** [comment]
I (partially based on the existing code for pools) implemented property management on both pools and datasets resources, and reading properties on both the resource and data versions of these. I ran into a bunch of intricacies/corner cases that I had not anticipated which complicated matters, but I'm pretty happy with the results. Some thoughts/notes in no particular order.

There are dataset and pool properties, and there are important differences between these. They are managed using different commands (`zfs` vs `zpool`) and use different flags during pool creation (`-O` vs `-o`). There are also user dataset properties which are somewhat free-form (though they are easy to recognize as they are always in the form of `namespace:name`), for which an empty string is a valid value.

Unsetting a property is nontrivial. For starters there is no equivalent of `zfs inherit` for pool properties. In addition some dataset properties don't have a _default_ value (even though they do have an _initial_ value), so they cannot be unset using `zfs inherit` either. Most pool properties start out with source `local`, so they seem to be in the same camp of having an _initial_ value but no _default_ value (which is probably why there is no `zpool inherit`). I've added logic to attempt to do the right thing for each property (or a warning if a property cannot be unset), but there might be edge cases I've missed.

Querying properties is also more complicated than it seems. Every pool has a dataset with the same name, and to get all properties of this pair you need to use both `zfs get` and `zpool get`. There are a few properties that are in both lists (`readonly`), but for the most part this doesn't matter as these always match. A _very important_ exception to this is `guid`; the `guid` returned from `zfs get guid` and `zpool get guid` **do not match**. I've continued to use the pool's `guid` (as was the case before) for pools.

Querying properties using `zfs/zpool get all` also doesn't actually return _all_ properties. Notably properties such as `userquota@username` will not be included unless specifically requested. Requesting both `all` and specific resources at the same time seems to not work (it behaves as if you only asked for `all`), so these need to be fetched separately. In cases where any properties that are used by the user aren't included in `get all` a second `get` is done to query the missing properties.

Some properties (sizes, dates, maybe others) are formatted by default when getting them. This is nice for human readable purposes, and in many cases will probably match the form used to set these values (I'd wager people are more likely to set the `quota` to `10G` than to `10737418240`), but there is some value in having the raw value so I get both.

For starters both forms are valid when setting, so we need to consider both to avoid terraform from persistently reporting a difference (though this cannot be avoided in all cases, detecting `1024M` is the same as `1G` would require a deeper understanding of what the properties represent than I think is desirable for the provider to have).

Having the raw value can also be useful if you want to do any computations with the properties (e.g. setting the `quota` of a pool to be 20% of the `quota` of its parent), though these are pretty niche. I've included the raw values in a separate calculated attribute `rawProperties`.

I've made no attempts to protect the user from invalid actions (e.g. changing an uneditable property such as `readonly` or `encrypted`) as just propagating the ZFS error message works just fine for this.

All in all it's a more involved process than I had hoped, but it seems to work pretty reliably from my testing. I've certainly not tried all properties (I don't even know what most of them do to be honest), but I've played around with most of the ones that seemed special in some way and have tried to pick something from each special combination/group that I noticed in the docs (e.g. all combinations of editable=yes/no inherit=yes/no that are visible in the docs).

File: internal/provider/resource_pool.go [hunk 1]
```diff
@@ -133,13 +167,13 @@ func resourcePoolCreate(ctx context.Context, d *schema.ResourceData, meta interf
 		}
 	}
 
-	vdev_spec, err := parseVdevSpecification(d.Get("mirror"), d.Get("devices"))
```
Comment (MaienM) [inline internal/provider/resource_pool.go:136]: This isn't really related, but looks like creating a pool using `device` blocks was broken, and I was trying to use that for testing.

File: internal/provider/zfs.go [hunk 2]
```diff
@@ -296,14 +538,19 @@ type CreatePool struct {
...
+		} else {
+			serialized_options += fmt.Sprintf(" -O '%s'='%s'", property, value)
+		}
 	}
 
+	callSshCommand(config, "zpool destroy %s", pool.name) // TODO remove
 	_, err := callSshCommand(config, "zpool create %s %s %s", serialized_options, pool.name, pool.vdevs)
 
 	if err != nil {
 		// We might have an error, but it's possible that the pool was still created
```
Comment (MaienM) [inline internal/provider/zfs.go:552]: I've left this intact as I wasn't sure what the motivation for this was, but as far as I can tell this isn't actually a useful action. The logic seems inverted (the fetched pool is returned _only_ if `describePool` returns an error) and the returned pool is never used. Maybe I'm missing something, or maybe the thing that this was used for has changed, but as things stand I think the whole "fetch-after-fail" could be removed entirely (also applies to `createDataset`).

File: internal/provider/resource_pool.go (hunk 1 above)
Comment (MathiasPius) [reply internal/provider/resource_pool.go:136]: Definitely an oversight, good catch. The terraform project I was using as a test only used mirrors, so I never hit that issue.

File: internal/provider/zfs.go (hunk 2 above)
Comment (MathiasPius) [reply internal/provider/zfs.go:552]: I understand the confusion, I believe I experienced I case wherein the pool would get created, but errors still occurred, such as with the underlying SSH connection for example, which means that the pool would exist, but errors still occurred.

The awkward error checking logic is to determine that case.

**Comment by MathiasPius** [comment]
This is amazing work, I really appreciate it! 

I think your reasoning is sound, and having both types of properties listed is a great compromise.

Amazing work!

**Review by MathiasPius — APPROVED** [review]

**Pull request merged**